### ISSUE TITLE
Multi classification on GPU runs fast only the first time

### ISSUE THREAD

**Issue by wingedskull** [dmlc/xgboost#6980]
Hello.
It is not an imagination, checked many times in R an Python, latest or one of the latest builds.
Using params like:
> multi:softmax, num_class=5, gpu_hist, gradient_based, eta=1.0 (for random forest) trees: 200 and boost rounds 15 
only the first run is fast. Trying to run this training even for second time (the same language or other) finishes it MUCH SLOWER,
What is even more weird, the time elapsed for next runs are high for any CPU/GPU. Much better Workstation shows the same or even slower execution compared to my cheap laptop. And!!!! MUCH SLOWER then CPU, for WS it is a special choque.
Three details more:
1) regression training runs with the same acceleration on GPU as it should. Doesn't matter form R or Python or C-API.
2) Being persistent (running code many-many times) finally leads to freezes in graphical interface (more then 15 sec before it became responsible again). I continued and everything hanged. Only power button helped.
3) We used latest Ubuntu and latest Win 10. XGB - 1.5.0 Your night builds and compiled from source code. All the same.

And the headshot, even twice:
1) I was playing with one OpenCL code, doing stupid things sometimes. But even after crashing or getting error messages all we back to normal for the next runs. It means that nVidia drivers are fine.
2) Checking library I saw cuda context creation, but NONE to destroy it after use.

Need solution ASAP, since we need to tune the model running it with many diff. params. And in such case each second counts.

**Comment by hcho3** [comment]
I am sorry to hear about your trouble. To help us identify the root cause, I will ask you a few questions.

Have you run into this issue in multiple machines? Or is it specific to your workstation? So far I have not run into this issue myself; our testing server runs the XGBoost code multiple times over and it does not freeze. 

Please post your training script so that we can run it ourselves and see if the problem occurs on our side. If not, the issue may be with the specific machines you are using (e.g. broken CUDA runtime or drivers).

**Comment by wingedskull** [comment]
Hello Philip
Thanks for an immediate answer. Sending you the code (Python for sake of compatibility) and data in zip.
Once more: GPU is engaged.
About run times (for you - to estimate the ratio). First Run: 35 sec, fastest of the next runs: 56.2 sec.
Switching to uniform and hist (core i5 - 4 cores + 4 hyperthreds) gives 14.5-15.5 sec
We are using Random Forest and boosted Random Forest params suggested in your docs.
VERY strange, since in other tests with reg:xxx models GPU calculations are much faster than CPU.







On Saturday, May 22, 2021, 1:25:39 AM GMT-5, Philip Hyunsu Cho ***@***.***> wrote: 







I am sorry to hear about your trouble. To help us identify the root cause, I will ask you a few questions.

Have you run into this issue in multiple machines? Or is it specific to your workstation? So far I have not run into this issue myself; our testing server runs the XGBoost code multiple times over and it does not freeze.

Please post your training script so that we can run it ourselves and see if the problem occurs on our side. If not, the issue may be with the specific machines you are using (e.g. broken CUDA runtime or drivers).

Speeds are the same in R and Python. And consistent for my laptop (i5, 10th gen, simple MX330 2Gb GDDR5) or workstation (do not have all specs on hands, but one of the latest Ryzen 7 and RTX2080). So, WS has the same time as my laptop for GPU!!!???

Dying from a desire to hear what happens.
And, I'd really appreciate if you will provide to me EXACT indications how to use in Python xgboost which I compiled locally (not through pip for stable build or nightly builds). Want to see what happens in this case. Please!!!!

**Comment by wingedskull** [comment]
[xg-test.zip](https://github.com/dmlc/xgboost/files/6527245/xg-test.zip)
Sending it here

**Comment by wingedskull** [comment]
Hello to the developers team.
Were you able to execute an example code several times, looking at how it works, including GPU/CPU?
For the moment is you could confirm the case - will be very important.
I just hope that it is really the issue when resources were allocated, context created but due to changes in trees some of them are not de-allocated where it should.
But also the presence of context creation - and no destruction in compiled library is worrying me.

**Comment by trivialfis** [comment]
I'm looking into them today.  Thanks for the patience.

**Comment by trivialfis** [comment]
Running it in a bash loop.  Will update the progress once it's finished.

**Comment by trivialfis** [comment]
Ran it 32 times with Python loop on CUDA 11, so far I haven't been able to reproduce it. Log and modified code are attached below.  The code changes are:

- "hist" -> "gpu_hist"
- Added for loop.

``` python
#!/usr/bin/env python
# coding: utf-8

# In[29]:
import pandas as pd
import numpy as np

# from collections import Counter
import os
import xgboost as xgb
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import time

# from ray import tune

dfs = pd.read_excel("input.xlsx", sheet_name="Sheet1", engine="openpyxl")

# In[188]:
X = dfs[
    [
        "f01",
        "f02",
        "f03",
        "f04",
        "f05",
        "f06",
        "f07",
        "f08",
        "f09",
        "f10",
        "f11",
        "f12",
        "f13",
    ]
]
y = dfs[["class"]]
Xtrain, Xtest, y_trv, y_tsv = train_test_split(X, y, test_size=0.25)
ytrv = np.array(y_trv)
for i in range(len(ytrv)):
    if ytrv[i] < 0:
        ytrv[i] = ytrv[i] + 2
    else:
        ytrv[i] = ytrv[i] + 1
ytsv = np.array(y_tsv)
for i in range(len(ytsv)):
    if ytsv[i] < 0:
        ytsv[i] = ytsv[i] + 2
    else:
        ytsv[i] = ytsv[i] + 1

# print(y_test['class'].values)
# print(Xtest.values)
dtrain = xgb.DMatrix(Xtrain.values, label=(ytrv))
dtest = xgb.DMatrix(Xtest.values, label=(ytsv))

# set xgboost params
param = {
    "eval_metric": "auc",
    "sampling_method": "uniform",  # or 'gradient_based', or 'uniform'
    "tree_method": "gpu_hist",  # or gpu_hist
    "colsample_bynode": 0.85,
    "subsample": 0.9,
    "nthread": 4,
    "min_child_weight": 3,
    "max_depth": 3,  # the maximum depth of each tree
    "eta": 1.0,  # the training step for each iteration
    "max_bin": 16,
    "num_parallel_tree": 200,
    "objective": "multi:softmax",  # error evaluation for multiclass training -
    "num_class": 5,
}  # the number of classes that exist in this datset

# num_round = 20  # the number of training iterations
# training and testing - numpy matrices

for i in range(32):
    t = time.process_time()
    bst = xgb.train(param, dtrain, num_boost_round=15)
    print(time.process_time() - t)

    preds = bst.predict(dtest)

    # evaluate predictions
    predictions = [round(value) for value in preds]
    accuracy = accuracy_score(ytsv, predictions)
    print("Accuracy: %.5f%%" % (accuracy * 100.0))
```
```
9.535810864
Accuracy: 67.14197%
9.145630353
Accuracy: 67.14197%
9.193478827
Accuracy: 67.14197%
9.127465213
Accuracy: 67.14197%
9.164171840999998
Accuracy: 67.14197%
9.176722685999998
Accuracy: 67.14197%
9.152670914000005
Accuracy: 67.14197%
9.147816884999997
Accuracy: 67.14197%
9.15949061900001
Accuracy: 67.14197%
9.164212014
Accuracy: 67.14197%
9.173735972000003
Accuracy: 67.14197%
9.201629443000002
Accuracy: 67.14197%
9.187759968000009
Accuracy: 67.14197%
9.144264339999992
Accuracy: 67.14197%
9.213034332000007
Accuracy: 67.14197%
9.146277424000004
Accuracy: 67.14197%
9.188005726
Accuracy: 67.14197%
9.221451680000001
Accuracy: 67.14197%
9.157122442999992
Accuracy: 67.14197%
9.131143097000006
Accuracy: 67.14197%
9.165762540999992
Accuracy: 67.14197%
9.147889808000002
Accuracy: 67.14197%
9.138431462
Accuracy: 67.14197%
9.151869046000002
Accuracy: 67.14197%
9.16471638600001
Accuracy: 67.14197%
9.137945572000007
Accuracy: 67.14197%
9.14844348699998
Accuracy: 67.14197%
9.154025891000003
Accuracy: 67.14197%
9.124639075000005
Accuracy: 67.14197%
9.136131904000024
Accuracy: 67.14197%
9.239540196000007
Accuracy: 67.14197%
9.260099731000025
Accuracy: 67.14197%
```

**Comment by wingedskull** [comment]
First thing to ask. Do you have nVidia GPU? If yes, which driver you are using?
Second thing: can you share the xgb build which you are using?
I want to check both things, since your machine definitely uses GPU correctly.

**Comment by wingedskull** [comment]
BTW.
Try, pls to use 'gradient_based' or explain me well why I should not use it?

**Comment by trivialfis** [comment]
> First thing to ask. Do you have nVidia GPU? If yes, which driver you are using?

Yes.  Ubuntu 20.04 with driver 460.73.01.  Also:
```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Wed_Jul_22_19:09:09_PDT_2020
Cuda compilation tools, release 11.0, V11.0.221
Build cuda_11.0_bu.TC445_37.28845127_0
```

> Second thing: can you share the xgb build which you are using?

The commit I used 7e846bb9652bb263414a49257fc6b9327ef848d3  since you mentioned you have compiled from source without listing the commit you are using.  I can try other commits too if you have specific one in mind.

> Try, pls to use 'gradient_based' or explain me well why I should not use it?

Just trying to run your script with minimum amount of changes, changed to gradient and running, right now the result looks pretty normal.

```
11.597548842999998
Accuracy: 68.00992%
11.407566096
Accuracy: 68.00992%
11.357898647999999
Accuracy: 68.00992%
11.376000288
Accuracy: 68.00992%
11.482708076000002
Accuracy: 68.00992%
11.413722822000004
Accuracy: 68.00992%
11.449749834000002
Accuracy: 68.00992%
11.433931326000007
Accuracy: 68.00992%
11.491348352999992
Accuracy: 68.00992%
11.474730394000005
Accuracy: 68.00992%
11.465334377999994
Accuracy: 68.00992%
11.446381126999995
Accuracy: 68.00992%
```

**Comment by wingedskull** [comment]
Important!
The same driver.
Made 2 things:
1) Pushed GPU from auto mode to performance mode always
2) Installed the latest night build for python
Executed under python 3: stable times, quite fast for my weak GPU
Tried python 2: stable times, but nearly 2 times slower... and a bit lower accuracy (1%) for the same code and conditions. 
Thus, started under python3 again. And guess what: execution times like for python 2!
So, to avoid thermal issues of laptop video card left it w/o any actions, to cool down a chip.
> Slow!
Closed terminal window and started it again in other one, with python3
> Slow!
Log off - OS GUI crashed! Text terminal, reboot
Computer restarted, the check in python3.
> Normal speed!

Any ideas?

**Comment by hcho3** [comment]
It appears that the issue is specific to your machine. Consider renting GPU-enabled machines from cloud services, such as Amazon EC2.

**Comment by wingedskull** [comment]
Jupiter uses Python 3 by default, so you never saw the issue.
I will use xbg under Python 3 only. But the issue is detected, so I highly recommend inserting the check for Python version, blocking GPU!!! under Python2

**Comment by hcho3** [comment]
XGBoost actually never supports Python 2. `pip install xgboost` will fail if you use Python 2.

**Comment by trivialfis** [comment]
I'm not entirely sure how you run xgboost under py2, there should be a lots of syntax errors ..

I will add the check, thanks for the suggestion!

**Comment by wingedskull** [comment]
I mean <python> and <python3> as a command. p3 - fine, just p - problems
